onnxruntime/core/providers/cuda/tensor/trilu_impl.cu

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "trilu_impl.h"
+#include <stdio.h>
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T, bool upper>
+__global__ void TriluKernel(
+    int64_t k,
+    const T* input_data,
+    T* output_data,
+    const CUDA_LONG N,
+    const fast_divmod batch_divmod_indices,
+    const fast_divmod row_col_divmod_indices) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+
+  int row, col;
+
+  row_col_divmod_indices.divmod(batch_divmod_indices.mod(id), row, col);
+  output_data[id] = ((((row + k) <= col) && upper) || (((row + k) >= col) && !upper)) ? input_data[id] : 0;
```

Comment:
I am not sure if we need to rewrite into following pattern so that the compiler can optimize the bool-type template parameter. 

if (upper) 
...
else
...